File: first_unique_char.py
#Assumption: the only characters are the ones contained on the ASCII table

class Solution:
    """
    @param str: str: the given string
    @return: char: the first unique character in a given string
    """
    def firstUniqChar(self, str):
        NUM_CHARS = 255
        # Write your code here
        count_and_first_pos = [[0,0] for i in range(0,255)]
        n = len(str)
        for i in range(n):
            ascii = ord(str[i])
            if (count_and_first_pos[ascii][0] == 0):
                count_and_first_pos[ascii][1] = i
            count_and_first_pos[ascii][0] += 1
        
        first_pos = n+1
        for element in count_and_first_pos:
            if element[0] == 1:
                if element[1] < first_pos:
                    first_pos = element[1]
        
        return str[first_pos]        

File: LintCode/first_unique_char.java
import java.util.*;

public class Solution {
    /**
     * @param str: str: the given string
     * @return: char: the first unique character in a given string
     */
    public char firstUniqChar(String str) {
        // Write your code here
        HashSet<char> set1 = new HashSet<char>();
        HashSet<char> set2 = new HashSet<char>();
        
        for (char c : str){
        	if (set2.contains(c)){
        		if (set1.contains(c)){
        			set1.remove(c);
        		}
        	} else {
        		set1.add(c);
        		set2.add(c);
        	}
        }
        Iterator<char> i = set1.iterator();
        return i.next();
    }
}